File: ios/HybridValidationTask.swift
import Foundation
import NitroModules

final class HybridValidationTask: HybridNativeValidationTaskSpec {
  private let stateMachine = TaskStateMachine()
  private let operationId: String
  private let session: ArchiveEngineSession
  private let engine: ArchiveEngine
  private let request: NativeValidationRequest

  private var progressCallback: ((NativeProgress) -> Void)?

  private var cachedState: String = "idle"
  private var cachedProgress: NativeProgress

  var state: String {
    cachedState
  }

  var progress: NativeProgress {
    cachedProgress
  }

  init(
    session: ArchiveEngineSession,
    engine: ArchiveEngine,
    request: NativeValidationRequest
  ) {
    let opId = UUID().uuidString
    self.operationId = opId
    self.session = session
    self.engine = engine
    self.request = request
    self.cachedProgress = NativeProgress(
      operationId: opId,
      phase: "preparing",
      processedBytes: 0,
      totalBytes: nil,
      processedEntries: 0,
      totalEntries: nil,
      currentEntry: nil,
      bytesPerSecond: nil,
      estimatedSecondsRemaining: nil,
      percentage: nil
    )
    super.init()
  }

  func start() throws -> Promise<NativeValidationResult> {
    Promise.async {
      try self.stateMachine.start()
      self.cachedState = "running"
      self.cachedProgress = NativeProgress(operationId: self.operationId, phase: "validating", processedBytes: 0, totalBytes: nil, processedEntries: 0, totalEntries: nil, currentEntry: nil, bytesPerSecond: nil, estimatedSecondsRemaining: nil, percentage: nil)

      let entries = self.session.entries
      let totalCount = entries.count
      let verifyChecksums = self.request.verifyChecksums ?? true
      let scanAllEntries = self.request.scanAllEntries ?? true
      let password = self.request.password

      self.emitProgress(phase: "validating", processedBytes: 0,
                         processedEntries: 0, currentEntry: nil,
                         totalEntries: totalCount)

      var issues: [NativeArchiveIssue] = []
      var checkedEntries = 0
      var checkedUncompressedBytes: UInt64 = 0
      var encryptedEntries = 0
      let startTime = CFAbsoluteTimeGetCurrent()
      let maxChecksumBytes: UInt64 = 64 * 1024 * 1024

      let entriesToValidate = scanAllEntries ? entries : entries.filter { $0.encrypted }

      for (index, entry) in entriesToValidate.enumerated() {
        try Task.checkCancellation()

        if entry.encrypted {
          encryptedEntries += 1
        }

        if verifyChecksums {
          if entry.encrypted && (password?.isEmpty ?? true) {
            issues.append(NativeArchiveIssue(
              code: "E_PASSWORD_REQUIRED",
              severity: "error",
              message: "Password is required to validate encrypted entry: \(entry.path)",
              entryPath: entry.path,
              entryIndex: Double(index)
            ))
            checkedEntries += 1
            continue
          }

          guard entry.uncompressedSize <= maxChecksumBytes else {
            issues.append(NativeArchiveIssue(
              code: "E_ARCHIVE_TOO_LARGE",
              severity: "warning",
              message: "Checksum validation skipped for large entry: \(entry.path)",
              entryPath: entry.path,
              entryIndex: Double(index)
            ))
            checkedEntries += 1
            continue
          }

          do {
            let maxBytes = max(entry.uncompressedSize, 1)
            let limit = min(maxBytes, maxChecksumBytes)
            _ = try await self.engine.readEntry(
              session: self.session,
              path: entry.path,
              limit: limit,
              password: password
            )
            checkedEntries += 1
            checkedUncompressedBytes += entry.uncompressedSize
          } catch let error as ArchiveDomainError {
            switch error {
            case .checksumMismatch:
              issues.append(NativeArchiveIssue(
                code: "E_CHECKSUM_MISMATCH",
                severity: "error",
                message: "Checksum validation failed for entry: \(entry.path)",
                entryPath: entry.path,
                entryIndex: Double(index)
              ))
            case .passwordRequired, .badPassword, .unsupportedEncryption:
              issues.append(NativeArchiveIssue(
                code: error.code,
                severity: "error",
                message: "\(error)",
                entryPath: entry.path,
                entryIndex: Double(index)
              ))
            case .bufferLimitExceeded, .entryNotFound:
              issues.append(NativeArchiveIssue(
                code: error.code,
                severity: "error",
                message: "\(error)",
                entryPath: entry.path,
                entryIndex: Double(index)
              ))
            default:
              throw error
            }
          }
        } else {
          checkedEntries += 1
          checkedUncompressedBytes += entry.uncompressedSize
        }

        self.emitProgress(phase: "validating", processedBytes: checkedUncompressedBytes,
                          processedEntries: index + 1, currentEntry: entry.path,
                          totalEntries: totalCount)
      }

      try self.stateMachine.succeed()
      self.cachedState = "succeeded"

      let duration = (CFAbsoluteTimeGetCurrent() - startTime) * 1000
      let isValid = !issues.contains { $0.severity == "error" }

      self.emitProgress(phase: "finalizing", processedBytes: checkedUncompressedBytes,
                        processedEntries: checkedEntries, currentEntry: nil,
                        totalEntries: totalCount, percentage: 100)

      return NativeValidationResult(
        operationId: self.operationId,
        valid: isValid,
        checkedEntries: Double(checkedEntries),
        checkedUncompressedBytes: checkedUncompressedBytes,
        encryptedEntries: Double(encryptedEntries),
        durationMs: duration,
        issues: issues
      )
    }
  }

  func cancel() throws -> Bool {
    let didCancel = stateMachine.cancel()
    if didCancel {
      cachedState = "cancelled"
      cachedProgress = NativeProgress(operationId: operationId, phase: "cleaning-up",
                                       processedBytes: 0, totalBytes: nil,
                                       processedEntries: 0, totalEntries: nil,
                                       currentEntry: nil, bytesPerSecond: nil,
                                       estimatedSecondsRemaining: nil, percentage: nil)
    }
    return didCancel
  }

  func onProgress(callback: @escaping (NativeProgress) -> Void) throws -> () -> Void {
    progressCallback = callback
    return { [weak self] in
      self?.progressCallback = nil
    }
  }

  private func emitProgress(
    phase: String,
    processedBytes: UInt64,
    processedEntries: Int,
    currentEntry: String?,
    totalEntries: Int? = nil,
    percentage: Double? = nil
  ) {
    let total = totalEntries.map(Double.init)
    let pct: Double?
    if let p = percentage {
      pct = p
    } else if let t = total, t > 0 {
      pct = min(Double(processedEntries) / t * 100, 100)
    } else {
      pct = nil
    }

    let snapshot = NativeProgress(
      operationId: operationId,
      phase: phase,
      processedBytes: processedBytes,
      totalBytes: nil,
      processedEntries: Double(processedEntries),
      totalEntries: total,
      currentEntry: currentEntry,
      bytesPerSecond: nil,
      estimatedSecondsRemaining: nil,
      percentage: pct
    )

    cachedProgress = snapshot
    progressCallback?(snapshot)
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/HybridValidationTask.kt
package com.mcodex.nitroarchive

import com.margelo.nitro.archive.*
import com.margelo.nitro.core.Promise
import com.mcodex.nitroarchive.application.ports.ArchiveEngine
import com.mcodex.nitroarchive.application.ports.ArchiveEngineSession
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.domain.TaskState
import com.mcodex.nitroarchive.domain.TaskStateMachine
import java.util.UUID
import java.util.zip.CRC32

class HybridValidationTask(
    private val session: ArchiveEngineSession,
    private val engine: ArchiveEngine,
    private val request: NativeValidationRequest
) : HybridNativeValidationTaskSpec() {

    private val stateMachine = TaskStateMachine()
    private val operationId = UUID.randomUUID().toString()
    private val listeners = mutableListOf<(NativeProgress) -> Unit>()
    private var latestProgress = NativeProgress(
        operationId = operationId,
        phase = "idle",
        processedBytes = 0L,
        totalBytes = null,
        processedEntries = 0.0,
        totalEntries = null,
        currentEntry = null,
        bytesPerSecond = null,
        estimatedSecondsRemaining = null,
        percentage = null
    )

    override val state: String get() = stateMachine.state.name.lowercase()
    override val progress: NativeProgress get() = latestProgress

    override fun start(): Promise<NativeValidationResult> = Promise.async {
        stateMachine.start()
        val startTime = System.currentTimeMillis()
        try {
            emitProgress(phase = "validating")

            val entries = session.entries
            val issues = mutableListOf<NativeArchiveIssue>()
            var checkedUncompressed = 0L
            var encryptedCount = 0
            val verifyChecksums = request.verifyChecksums ?: true
            val scanAll = request.scanAllEntries ?: true
            val maxChecksumBytes = 64L * 1024L * 1024L
            val entriesToValidate = if (scanAll) entries else entries.filter { it.encrypted }

            emitProgress(
                phase = "validating",
                totalEntries = entriesToValidate.size.toDouble()
            )

            for ((i, entry) in entriesToValidate.withIndex()) {
                if (stateMachine.state == TaskState.CANCELLING) {
                    throw ArchiveDomainError.OperationCancelled()
                }

                if (entry.encrypted) {
                    encryptedCount++
                }

                checkedUncompressed += entry.uncompressedSize

                emitProgress(
                    phase = "validating",
                    processedEntries = i.toDouble(),
                    totalEntries = entriesToValidate.size.toDouble(),
                    currentEntry = entry.path
                )

                if (entry.kind == "file" && (verifyChecksums || scanAll)) {
                    if (entry.encrypted && request.password.isNullOrEmpty()) {
                        issues.add(NativeArchiveIssue(
                            code = "E_PASSWORD_REQUIRED",
                            severity = "error",
                            message = "Password is required to validate encrypted entry: ${entry.path}",
                            entryPath = entry.path,
                            entryIndex = entry.index.toDouble()
                        ))
                        continue
                    }
                    if (entry.uncompressedSize > maxChecksumBytes) {
                        issues.add(NativeArchiveIssue(
                            code = "E_ARCHIVE_TOO_LARGE",
                            severity = "warning",
                            message = "Checksum validation skipped for large entry: ${entry.path}",
                            entryPath = entry.path,
                            entryIndex = entry.index.toDouble()
                        ))
                        continue
                    }
                    try {
                        val data = engine.readEntry(
                            session = session,
                            path = entry.path,
                            limit = entry.uncompressedSize.coerceAtLeast(1),
                            password = request.password
                        )

                        if (verifyChecksums && entry.crc32 != null && data.isNotEmpty()) {
                            val actualCrc = crc32(data)
                            if (actualCrc != entry.crc32) {
                                issues.add(NativeArchiveIssue(
                                    code = "E_CHECKSUM_MISMATCH",
                                    severity = "error",
                                    message = "CRC mismatch for entry: ${entry.path}",
                                    entryPath = entry.path,
                                    entryIndex = entry.index.toDouble()
                                ))
                            }
                        }
                    } catch (e: ArchiveDomainError.ChecksumMismatch) {
                        issues.add(NativeArchiveIssue(
                            code = "E_CHECKSUM_MISMATCH",
                            severity = "error",
                            message = e.message ?: "CRC mismatch",
                            entryPath = entry.path,
                            entryIndex = entry.index.toDouble()
                        ))
                    } catch (e: ArchiveDomainError.BadPassword) {
                        issues.add(NativeArchiveIssue(
                            code = e.code,
                            severity = "error",
                            message = e.message ?: "Incorrect password",
                            entryPath = entry.path,
                            entryIndex = entry.index.toDouble()
                        ))
                    } catch (e: ArchiveDomainError.UnsupportedEncryption) {
                        issues.add(NativeArchiveIssue(
                            code = e.code,
                            severity = "error",
                            message = e.message ?: "Unsupported encryption",
                            entryPath = entry.path,
                            entryIndex = entry.index.toDouble()
                        ))
                    } catch (e: ArchiveDomainError.OperationCancelled) {
                        throw e
                    } catch (e: Exception) {
                        if (scanAll) {
                            issues.add(NativeArchiveIssue(
                                code = "E_IO",
                                severity = "error",
                                message = "Failed to read entry '${entry.path}': ${e.message}",
                                entryPath = entry.path,
                                entryIndex = entry.index.toDouble()
                            ))
                        } else {
                            throw e
                        }
                    }
                }
            }

            emitProgress(
                phase = "finalizing",
                processedEntries = entriesToValidate.size.toDouble(),
                totalEntries = entriesToValidate.size.toDouble()
            )

            stateMachine.succeed()

            NativeValidationResult(
                operationId = operationId,
                valid = issues.isEmpty(),
                checkedEntries = entriesToValidate.size.toDouble(),
                checkedUncompressedBytes = checkedUncompressed,
                encryptedEntries = encryptedCount.toDouble(),
                durationMs = (System.currentTimeMillis() - startTime).toDouble(),
                issues = issues.toTypedArray()
            )
        } catch (e: ArchiveDomainError.OperationCancelled) {
            stateMachine.markCancelled()
            throw e
        } catch (e: ArchiveDomainError) {
            stateMachine.fail()
            throw e
        } catch (e: Exception) {
            stateMachine.fail()
            throw e
        }
    }

    override fun cancel(): Boolean = stateMachine.cancel()

    override fun onProgress(callback: (progress: NativeProgress) -> Unit): () -> Unit {
        listeners.add(callback)
        return { listeners.remove(callback) }
    }

    private fun emitProgress(
        phase: String,
        processedBytes: Long = 0L,
        totalBytes: Long? = null,
        processedEntries: Double = 0.0,
        totalEntries: Double? = null,
        currentEntry: String? = null
    ) {
        val percentage = when {
            totalBytes != null && totalBytes > 0 ->
                minOf(processedBytes.toDouble() / totalBytes.toDouble() * 100.0, 100.0)
            totalEntries != null && totalEntries > 0 ->
                minOf(processedEntries / totalEntries * 100.0, 100.0)
            else -> null
        }
        val snapshot = NativeProgress(
            operationId = operationId,
            phase = phase,
            processedBytes = processedBytes,
            totalBytes = totalBytes,
            processedEntries = processedEntries,
            totalEntries = totalEntries,
            currentEntry = currentEntry,
            bytesPerSecond = null,
            estimatedSecondsRemaining = null,
            percentage = percentage
        )
        latestProgress = snapshot
        listeners.forEach { it(snapshot) }
    }

    companion object {
        fun crc32(data: ByteArray): UInt {
            val crc = CRC32()
            crc.update(data)
            return crc.value.toUInt()
        }
    }
}
